internal/util/update/resource-merge.go

```diff
@@ -64,165 +66,309 @@ func (u ResourceMergeUpdater) Update(options UpdateOptions) error {
 	if err != nil {
 		return err
 	}
 
-	if err := MergeSubPackages(options.PackagePath, updated.AbsPath(), original.AbsPath()); err != nil {
-		return err
+	return nil
+}
+
+// findAllSubpackages traverses the packages in local, updated and original
+// and finds all subpackages. A subpackage is a subdirectory underneath the
+// root that has a Kptfile in it.
+// The list is sorted in increasing order based on the depth of the subpackage
+// relative to the root package.
+func findAllSubpackages(local, updated, original string) ([]string, error) {
+	uniquePaths := make(map[string]bool)
+	for _, path := range []string{local, updated, original} {
+		paths, err := pathutil.DirsWithFile(path, kptfile.KptFileName, true)
+		if err != nil {
+			return []string{}, err
+		}
+		for _, p := range paths {
+			relPath, err := filepath.Rel(path, p)
+			if err != nil {
+				return []string{}, err
+			}
+			uniquePaths[relPath] = true
+		}
+	}
+	var paths []string
+	for p := range uniquePaths {
+		paths = append(paths, p)
 	}
+	sort.Slice(paths, func(i, j int) bool {
+		iPath := paths[i]
+		jPath := paths[j]
+		if iPath == "." {
+			return true
+		}
+		if jPath == "." {
+			return false
+		}
+		iSegmentCount := len(strings.Split(iPath, "/"))
+		jSegmentCount := len(strings.Split(jPath, "/"))
+		return iSegmentCount < jSegmentCount
+	})
+	return paths, nil
+}
 
-	// merge the Resources: original + updated + dest => dest
-	err = filters.Merge3{
-		OriginalPath: original.AbsPath(),
-		UpdatedPath:  updated.AbsPath(),
-		DestPath:     options.PackagePath,
-		// TODO: Write a test to ensure this is set
-		MergeOnPath: true,
-	}.Merge()
+// updatePackage updates the package in the location specified by localPath
+// using the provided paths to the updated version of the package and the
+// original version of the package.
+func (u ResourceMergeUpdater) updatePackage(subPkgPath, localPath, updatedPath, originalPath string) error {
+	localExists, err := exists(localPath)
 	if err != nil {
 		return err
 	}
 
-	return ReplaceNonKRMFiles(updated.AbsPath(), original.AbsPath(), options.PackagePath)
-}
-
-// updatedKptfile returns a Kptfile to replace the existing local Kptfile as part of the update
-func (u ResourceMergeUpdater) updatedKptfile(updatedPath, originalPath string, options UpdateOptions) (
-	kptfile.KptFile, error) {
-	cmd := exec.Command("git", "rev-parse", "--verify", "HEAD")
-	cmd.Dir = updatedPath
-	cmd.Env = os.Environ()
-	cmd.Stderr = os.Stderr
-	b, err := cmd.Output()
+	updatedExists, err := exists(updatedPath)
 	if err != nil {
-		return kptfile.KptFile{}, err
+		return err
 	}
-	commit := strings.TrimSpace(string(b))
-	updatedKf, err := kptfileutil.ReadFile(updatedPath)
+
+	originalExists, err := exists(originalPath)
 	if err != nil {
-		updatedKf, err = kptfileutil.ReadFile(options.PackagePath)
+		return err
+	}
+
+	switch {
+	// Check if subpackage has been added both in upstream and in local
```

Comment:
I see what you are doing here. You are addressing various cases of update scenarios at package level. My argument is, if we address the same for single package scenario at file level, we need not have package level logic. Handling package level scenarios might not address file level scenarios. 
For example, what should be the expected behavior if a file with same name is added in both upstream and local in single package scenario. If we address this case then the above case is automatically addressed.

Reply:
So I think handling packages separately from filers/folders are better for a couple of reasons:
 * This means that the implementation match how we conceptually think about updating packages. This also makes the implementation easier since we no longer need to handle Kptfiles and arbitrary depths in a folder structure. Kptfiles will always be merged within the context of the package to which they belong.
 * There are scenarios where we should do things differently when we are dealing with packages instead of folders/files. We know we need a way to delete dependencies that have been removed from upstream, and while we have had discussions on different ways to handle this, it will require logic that is different from regular folders.
 * This gives us flexibility long-term to handle packages differently and it allows us to provide clear feedback to users about the cause of any conflicts. An example would be that we can tell if a conflict is due to conflicts between local subpackages vs conflicts with dependencies.

This doesn't mean we shouldn't try to avoid duplicate code, so we should still aim to reuse code and functionality between packages and files/folders.

Reply:
@mortent I agree with the idea. Is the PR work in progress ? Are you planning to add anything more? This PR looks good to me and we can iterate on top of it.